### PR TITLE
Avoid inconsistent when send eos quickly

### PR DESCRIPTION
**Pull request by Pslydhh**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

pre-condition: use pipeline and set pipeline_dop = 1
sql: **SELECT 1 EXCEPT SELECT 1;**
This query maybe never done, and pipeline's thread execute forever with has_output_for_pipeline.

## Problem Summary(Required) ：

**When query like "SELECT 1 EXCEPT SELECT 1", and we use pipeline engine with pipeline_dop = 1.**

ExchangeSinkOperator will send **2 requests**, The first request is one line data '1' **with is_pipeline_level_shuffle = false**, second request is **Eos** with **is_pipeline_level_shuffle = true**, So this lead to a corner case with **chunk_queue isn't empty**, **is_pipeline_level_shuffle = true** and **_has_chunks_per_driver_sequence[item.driver_sequence]==false**, so we can't get data through ExchangeSourceOperator's has_output.

Just set **has_chunks_per_driver_sequence[item.driver_sequence] = true** when push data into _chunk_queue, to fix this.

### CODE REVIEW

File: be/src/runtime/data_stream_recvr.cc [hunk 1]
```diff
@@ -520,6 +520,8 @@ Status DataStreamRecvr::SenderQueue::add_chunks(const PTransmitChunkParams& requ
                     _short_circuit_driver_sequences.end()) {
                     continue;
                 }
+            }
+            if (item.driver_sequence >= 0) {
                 _has_chunks_per_driver_sequence[item.driver_sequence] = true;
             }
```
Comment (liuyehcf) [inline be/src/runtime/data_stream_recvr.cc:526]: Why set driver sequence when pipeline level shuffle disabled, is there any other plans?

Comment (Pslydhh) [reply be/src/runtime/data_stream_recvr.cc:526]: Because second request maybe change _is_pipeline_level_shuffle.

**Comment by kangkaisen** [comment]
> The first request is one line data '1' with is_pipeline_level_shuffle = false, second request is Eos with is_pipeline_level_shuffle = true

I don‘t understand it.

**Comment by liuyehcf** [comment]
The root cause of the problem is as follows

* ExchangeSink's partition type is `HASH_PARTITION`
* at sender side, be will actually send packets in `UNPARTITIONED`  way because dop is 1, and in this way `pipeline level shuffle` is not performed
    ```cpp
    Status ExchangeSinkOperator::push_chunk(RuntimeState* state, const vectorized::ChunkPtr& chunk) {
    uint16_t num_rows = chunk->num_rows();
    if (num_rows == 0) {
        return Status::OK();
    }
    if (_part_type == TPartitionType::UNPARTITIONED || (_num_shuffles == 1 && _channels.size() == 1)) {
        if (_chunk_request == nullptr) {
            _chunk_request = std::make_shared<PTransmitChunkParams>();
        }
    ```
* at receiver side, be still expected to receive package in `pipeline level shuffle` way.
* The mismatch of sender side and receiver side leads to this problem

How to fix it? Change the condition in `FragmentExecutor::_decompose_data_sink_to_operator` as follows, i.e. only set `is_pipeline_level_shuffle` to true only if `dest_dop > 1`

```cpp
        if (sender->get_partition_type() == TPartitionType::HASH_PARTITIONED ||
            sender->get_partition_type() == TPartitionType::BUCKET_SHUFFLE_HASH_PARTITIONED) {
            dest_dop = t_stream_sink.dest_dop;
            if (dest_dop > 1) {
                is_pipeline_level_shuffle = true;
            }
```

**Comment by Pslydhh** [comment]
> > The first request is one line data '1' with is_pipeline_level_shuffle = false, second request is Eos with is_pipeline_level_shuffle = true
> 
> I don‘t understand it.
image

"select 1" caused ExchangeSink send two requests: **first is data '1', second is eos;**
and **is_pipeline_level_shuffle** is different in there two requests, so lead to inconsistent.
This is a mistake caused by details.

**Review by liuyehcf — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 2]
```diff
@@ -347,8 +347,10 @@ void FragmentExecutor::_decompose_data_sink_to_operator(RuntimeState* runtime_st
         int32_t dest_dop = -1;
         if (sender->get_partition_type() == TPartitionType::HASH_PARTITIONED ||
             sender->get_partition_type() == TPartitionType::BUCKET_SHUFFLE_HASH_PARTITIONED) {
-            is_pipeline_level_shuffle = true;
             dest_dop = t_stream_sink.dest_dop;
```
Comment (satanson) [inline be/src/exec/pipeline/fragment_executor.cpp:350]: add some comment, hard to understand

Comment (Pslydhh) [reply be/src/exec/pipeline/fragment_executor.cpp:350]: Commented

**Review by satanson — DISMISSED** [review]

**Review by satanson — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**